plugins/teststeps/teststeps.go

```diff
@@ -27,49 +27,46 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
+	type tgtErr struct {
+		target *target.Target
+		err error
+	}
+	tgtErrCh := make(chan tgtErr)
 	for {
 		select {
-		case target := <-ch.In:
-			if target == nil {
+		case tgt := <-ch.In:
```

Comment:
Could it be that ForEachTarget has read nil from the In channel _before_ all the targets processed? It would lead to leaking the targets..